tests/fixtures/source/blocks/references/file.rst

```diff
@@ -0,0 +1,2 @@
+
+:file:`Link name::Symfony/Component/Notifier/Bridge/AmazonSns/README.md`
```

Comment:
Are you agree with this syntax ?

Do you have a better idea ?

Reply:
As we link to GitHub code, I propose to have sth like:

`:file:symfony/symfony::src/......`

so we could also link to: `symfony/symfony-docs` repo too.

Maybe we should rename it anyway to `:github-file:` ?

Reply:
We want inject `symfonyVersion` for `symfony` and `symfony-docs` not other repository.

I suggest

``:github-file:`My Link::symfony::src/` `` produce `https://github.com/symfony/symfony/blob/6.3/src`

``:github-file:`My Link::symfony-docs::src/` `` produce `https://github.com/symfony/symfony-docs/blob/6.3/src`

``:github-file:`My Link::foo/bar::src/` `` produce `https://github.com/foo/bar/src`

This way, it's easy to configure for usual links and we can also configura any github paths

WDYT ?

Reply:
If others agree I like it 😃

Reply:
If we follow the traditional RST syntax:

```rst
:ref:`The Label <the-reference-id>`
:doc:`The Label </the/doc/path>`
```

This could be instead:

```rst
:github:`My Link <symfony/symfony:src/>`
:github:`My Link <symfony/symfony-docs:src/>`
:github:`My Link <foo/bar:src/>`
```

Comments:

* `:github-file:` could be shortened as `:github:`. If in the future we add support for linking to issues or PR, we could create `:github-issue:` but keep the short `:github:` for the most common case.
* I wouldn't add a special case for the `symfony/*` repos because that always increases the learning curve: What repo is `symfony` ?¡can't happen if we use the explicit `symfony/symfony`.
* About the separator between repo name and path, we could use just `:` instead of `::` ... because `:` can't happen in repo names or file paths.

In any case, I'm still not 100% sure about this. I know we're trying to solve a real problem we have, but creating many custom RST tags would create a meta-language based on RST, which complicates maintenance and docs contributions.

But let's hear the opinions from @xabbuh, @wouterj and @weaverryan. Thanks!

Reply:
I understand your Problem, but that's exactly where the docs-builder can ease work.

I like and would go with:
```rst
:github:`My Link <symfony/symfony:src/>`
```

Reply:
I pushed a new version. If we don't want magic, branch must be explicit, we can use `{symfonyVersion}` special placeholder for symfony repos.

```rst
:github:`First link ::symfony/symfony:4.0/src/Symfony/Component/Notifier/Bridge/AmazonSns/README.md::`

:github:`Link with branch substitution ::foo/bar:{symfonyVersion}/src/baz::`
```
It seems complicated to use following syntax (I love it !)
```rst
`:github:`Link with branch substitution <foo/bar:{symfonyVersion}/src/baz>`
```
Indeed `rst-parser` consider it as a real `Doctrine\RST\Span\SpanToken` and value sent to our code (`$spanToken->get('url')`) is only `foo/bar:{symfonyVersion}/src/baz` (without link name) not `Link with branch substitution <foo/bar:{symfonyVersion}/src/baz>`

Reply:
The other day we merged https://github.com/symfony/symfony-docs/pull/17967 and seeing these "URI templates", I come up with an alternative idea to solve this problem:

(1) What we want to do: link to some GitHub file, but use a different version for each doc branch:

https://github.com/symfony/amazon-sns-notifier/blob/5.4/README.md
https://github.com/symfony/amazon-sns-notifier/blob/6.1/README.md
https://github.com/symfony/amazon-sns-notifier/blob/6.2/README.md
https://github.com/symfony/amazon-sns-notifier/blob/6.3/README.md

(2) What if we use "URI templates" for this kind of links?

Instead of creating a new custom directive:

```rst
:github:`README file <symfony/amazon-sns-notifier:README.md/>`
:github:`README file <symfony/amazon-sns-notifier:{6.2}/README.md/>`
```

We could use the normal RST links but with variables inside the URL:

```rst
`README file`_

.. _`README file`: https://github.com/symfony/amazon-sns-notifier/blob/{version}/README.md

.. OR, embedded:

`README file <https://github.com/symfony/amazon-sns-notifier/blob/{version}/README.md>`_
```

-----

The advantages that I see to this approach is that:

* Docs are still very readable
* Maintenance is the same as before, not more complicated
* There's no need to create a custom directive

The disadvantages:

* We need to update `docs-builder` to process the links and replace `{version}` by the current Symfony version of the docs being parsed
* We need to be careful about the URI variable pattern; it needs to be `{version}`, not `{ version }`, etc.

What do you think?

Reply:
Makes sense 👍 

> We need to be careful about the URI variable pattern; it needs to be {version}, not { version }, etc.

Can be ensured by DOCtor-RST 👍

Reply:
Thanks for your feedback. Native link with `{version}` substitution is the most readable indeed and easy to use.

I pushed new version in this direction.

Hope is ok :pray:

(I will create DOCtor-RST rule when this PR will be merge to ensure good usage)